File: src/golay.ts
// sdk/src/golay.ts
// 🔱 Golay(24,12) Extended Binary Golay Code (RE-SURGERY)
// Corrects up to 3 bit errors, detects 4 bit errors.

// Standard Golay Parity Matrix P (12x12) - Systematic Form [I12 | P]
// Every row has weight 7.
// Verified: P*P^T = I (mod 2), all rows weight 7.
// Derived from g(x) = x^11+x^10+x^6+x^5+x^4+x^2+1 with parity extension.
const P = [
    0xC75, 0x49F, 0xD4B, 0x6E3, 0x9B3, 0xB66,
    0xECC, 0x1ED, 0x3DA, 0x7B4, 0xB1D, 0xE3A
];

// R7-P0-1: P^T (transpose of P) — required for second syndrome computation.
// P is NOT symmetric, so s2 = s*P^T must use this separate matrix.
// Without this, decodeWord fails ~24% of weight-2 and ~50% of weight-3 error patterns.
const PT = [
    0xAE3, 0xF25, 0x16F, 0x2DE, 0x5BC, 0xB78,
    0x9D5, 0xC8F, 0x63B, 0xC76, 0x7C9, 0xF92
];

const UNIT_VECTORS = [
    0x800, 0x400, 0x200, 0x100, 0x080, 0x040,
    0x020, 0x010, 0x008, 0x004, 0x002, 0x001
];

/**
 * Calculates weight (number of set bits).
 */
function weight(n: number): number {
    let count = 0;
    let temp = n & 0xFFF;
    while (temp > 0) {
        temp &= (temp - 1);
        count++;
    }
    return count;
}

/**
 * Multiplies a 12-bit vector by the parity matrix P.
 */
function multiplyP(v: number): number {
    let res = 0;
    for (let i = 0; i < 12; i++) {
        if ((v >> (11 - i)) & 1) {
            res ^= P[i];
        }
    }
    return res & 0xFFF;
}

/**
 * R7-P0-1: Multiplies a 12-bit vector by P^T (transpose of P).
 * Required for second syndrome computation in decodeWord steps 3-5.
 */
function multiplyPT(v: number): number {
    let res = 0;
    for (let i = 0; i < 12; i++) {
        if ((v >> (11 - i)) & 1) {
            res ^= PT[i];
        }
    }
    return res & 0xFFF;
}

function encodeWord(msg: number): number {
    const parity = multiplyP(msg & 0xFFF);
    return ((msg & 0xFFF) << 12) | parity;
}

/**
 * Full Syndrome Decoding for Golay(24,12)
 */
function decodeWord(received: number): { msg: number; corrected: number; uncorrectable: boolean } {
    let r_m = (received >> 12) & 0xFFF;
    const r_p = received & 0xFFF;
    
    // Syndrome s = r_m * P + r_p
    const s = multiplyP(r_m) ^ r_p;
    
    if (s === 0) return { msg: r_m, corrected: 0, uncorrectable: false };

    // 1. wt(s) <= 3 -> Error in parity
    if (weight(s) <= 3) {
        return { msg: r_m, corrected: weight(s), uncorrectable: false };
    }

    // 2. wt(s + P_i) <= 2 -> Error in msg bit i (+ possible 1-2 errors in parity)
    for (let i = 0; i < 12; i++) {
        if (weight(s ^ P[i]) <= 2) {
            return { msg: r_m ^ UNIT_VECTORS[i], corrected: weight(s ^ P[i]) + 1, uncorrectable: false };
        }
    }

    // 3. R7-P0-1: Second syndrome s2 = s * P^T (NOT s * P — P is not symmetric!)
    const s2 = multiplyPT(s);

    // 4. wt(s2) <= 3 -> Error in message
    if (weight(s2) <= 3) {
        return { msg: r_m ^ s2, corrected: weight(s2), uncorrectable: false };
    }

    // 5. wt(s2 + PT_i) <= 2 -> Error in parity bit i + message errors
    // R7-P0-1: Must use PT rows (columns of P), not P rows
    for (let i = 0; i < 12; i++) {
        if (weight(s2 ^ PT[i]) <= 2) {
            const error_m = s2 ^ PT[i];
            return { msg: r_m ^ error_m, corrected: weight(s2 ^ PT[i]) + 1, uncorrectable: false };
        }
    }

    return { msg: r_m, corrected: 0, uncorrectable: true };
}

export function golayEncode(data: Uint8Array): Uint8Array {
    // B1-6: Correct output size calculation to avoid buffer overflow
    const out = new Uint8Array(Math.ceil(data.length / 3) * 6);
    let outIdx = 0;
    
    // Process in 3-byte blocks -> two 12-bit words -> two 24-bit (3-byte) codewords
    for (let i = 0; i < data.length; i += 3) {
        const b1 = data[i];
        const b2 = i + 1 < data.length ? data[i + 1] : 0;
        const b3 = i + 2 < data.length ? data[i + 2] : 0;

        // Word 1 (12 bits)
        const w1 = (b1 << 4) | (b2 >> 4);
        const c1 = encodeWord(w1);
        out[outIdx++] = (c1 >> 16) & 0xFF;
        out[outIdx++] = (c1 >> 8) & 0xFF;
        out[outIdx++] = c1 & 0xFF;

        if (i + 1 < data.length) {
            // Word 2 (12 bits)
            const w2 = ((b2 & 0x0F) << 8) | b3;
            const c2 = encodeWord(w2);
            out[outIdx++] = (c2 >> 16) & 0xFF;
            out[outIdx++] = (c2 >> 8) & 0xFF;
            out[outIdx++] = c2 & 0xFF;
        }
    }
    return out;
}

export function golayDecode(encoded: Uint8Array): { data: Uint8Array; corrected: number; uncorrectable: boolean } {
    // B1-6: Throws if length is not a multiple of 6
    if (encoded.length % 6 !== 0) {
        throw new Error("Invalid Golay encoded data: length must be multiple of 6");
    }
    const outLen = Math.floor(encoded.length / 2);
    const out = new Uint8Array(outLen);
    let outIdx = 0;
    let totalCorrected = 0;
    let anyUncorrectable = false;

    for (let i = 0; i < encoded.length; i += 6) {
        const c1 = (encoded[i] << 16) | (encoded[i + 1] << 8) | encoded[i + 2];
        const res1 = decodeWord(c1);
        totalCorrected += res1.corrected;
        if (res1.uncorrectable) anyUncorrectable = true;

        if (outIdx < outLen) out[outIdx++] = (res1.msg >> 4) & 0xFF;
        const b2_high = (res1.msg & 0x0F) << 4;

        if (i + 3 < encoded.length) {
            const c2 = (encoded[i + 3] << 16) | (encoded[i + 4] << 8) | encoded[i + 5];
            const res2 = decodeWord(c2);
            totalCorrected += res2.corrected;
            if (res2.uncorrectable) anyUncorrectable = true;

            if (outIdx < outLen) out[outIdx++] = b2_high | (res2.msg >> 8);
            if (outIdx < outLen) out[outIdx++] = res2.msg & 0xFF;
        } else {
            if (outIdx < outLen) out[outIdx++] = b2_high;
        }
    }

    return { data: out, corrected: totalCorrected, uncorrectable: anyUncorrectable };
}

File: wasm/grg-wasm/src/golay.rs
//! Golay(24,12) Extended Binary Code
//! Corrects up to 3 bit errors, detects 4 bit errors.
//! Direct port from OpenTTT golay.ts — R7-P0-1 P^T fix included.

/// Standard Golay Parity Matrix P (12x12)
const P: [u16; 12] = [
    0xC75, 0x49F, 0xD4B, 0x6E3, 0x9B3, 0xB66,
    0xECC, 0x1ED, 0x3DA, 0x7B4, 0xB1D, 0xE3A,
];

/// P^T — transpose of P (required for correct syndrome decoding)
const PT: [u16; 12] = [
    0xAE3, 0xF25, 0x16F, 0x2DE, 0x5BC, 0xB78,
    0x9D5, 0xC8F, 0x63B, 0xC76, 0x7C9, 0xF92,
];

const UNIT: [u16; 12] = [
    0x800, 0x400, 0x200, 0x100, 0x080, 0x040,
    0x020, 0x010, 0x008, 0x004, 0x002, 0x001,
];

#[inline]
fn weight(n: u16) -> u32 { (n & 0xFFF).count_ones() }

#[inline]
fn multiply_p(v: u16) -> u16 {
    let mut res = 0u16;
    for i in 0..12u16 {
        if (v >> (11 - i)) & 1 == 1 { res ^= P[i as usize]; }
    }
    res & 0xFFF
}

#[inline]
fn multiply_pt(v: u16) -> u16 {
    let mut res = 0u16;
    for i in 0..12u16 {
        if (v >> (11 - i)) & 1 == 1 { res ^= PT[i as usize]; }
    }
    res & 0xFFF
}

fn encode_word(msg: u16) -> u32 {
    let parity = multiply_p(msg & 0xFFF);
    ((msg as u32 & 0xFFF) << 12) | parity as u32
}

pub struct DecodeResult {
    pub msg: u16,
    pub corrected: u32,
    pub uncorrectable: bool,
}

fn decode_word(received: u32) -> DecodeResult {
    let r_m = ((received >> 12) & 0xFFF) as u16;
    let r_p = (received & 0xFFF) as u16;
    let s = multiply_p(r_m) ^ r_p;

    if s == 0 {
        return DecodeResult { msg: r_m, corrected: 0, uncorrectable: false };
    }
    // Step 1: wt(s) <= 3
    if weight(s) <= 3 {
        return DecodeResult { msg: r_m, corrected: weight(s), uncorrectable: false };
    }
    // Step 2: wt(s ^ P[i]) <= 2
    for i in 0..12 {
        let sp = s ^ P[i];
        if weight(sp) <= 2 {
            return DecodeResult { msg: r_m ^ UNIT[i], corrected: weight(sp) + 1, uncorrectable: false };
        }
    }
    // Step 3: second syndrome (R7-P0-1 fix: use P^T)
    let s2 = multiply_pt(s);
    if weight(s2) <= 3 {
        return DecodeResult { msg: r_m ^ s2, corrected: weight(s2), uncorrectable: false };
    }
    // Step 4: wt(s2 ^ PT[i]) <= 2
    for i in 0..12 {
        let s2p = s2 ^ PT[i];
        if weight(s2p) <= 2 {
            return DecodeResult { msg: r_m ^ s2p, corrected: weight(s2p) + 1, uncorrectable: false };
        }
    }
    DecodeResult { msg: r_m, corrected: 0, uncorrectable: true }
}

/// Encode bytes using Golay(24,12)
pub fn golay_encode(data: &[u8]) -> Vec<u8> {
    let out_len = ((data.len() + 2) / 3) * 6;
    let mut out = vec![0u8; out_len];
    let mut out_idx = 0;

    let mut i = 0;
    while i < data.len() {
        let b1 = data[i] as u32;
        let b2 = if i + 1 < data.len() { data[i + 1] as u32 } else { 0 };
        let b3 = if i + 2 < data.len() { data[i + 2] as u32 } else { 0 };

        let w1 = ((b1 << 4) | (b2 >> 4)) as u16;
        let c1 = encode_word(w1);
        out[out_idx]     = ((c1 >> 16) & 0xFF) as u8;
        out[out_idx + 1] = ((c1 >> 8)  & 0xFF) as u8;
        out[out_idx + 2] = (c1         & 0xFF) as u8;
        out_idx += 3;

        if i + 1 < data.len() {
            let w2 = (((b2 & 0x0F) << 8) | b3) as u16;
            let c2 = encode_word(w2);
            out[out_idx]     = ((c2 >> 16) & 0xFF) as u8;
            out[out_idx + 1] = ((c2 >> 8)  & 0xFF) as u8;
            out[out_idx + 2] = (c2         & 0xFF) as u8;
            out_idx += 3;
        }
        i += 3;
    }
    out[..out_idx].to_vec()
}

pub struct GolayDecoded {
    pub data: Vec<u8>,
    pub corrected: u32,
    pub uncorrectable: bool,
}

/// Decode Golay(24,12) encoded bytes
pub fn golay_decode(encoded: &[u8]) -> Result<GolayDecoded, &'static str> {
    if encoded.len() % 6 != 0 {
        return Err("Golay encoded length must be multiple of 6");
    }
    let out_len = encoded.len() / 2;
    let mut out = vec![0u8; out_len];
    let mut out_idx = 0;
    let mut total_corrected = 0u32;
    let mut any_uncorrectable = false;

    let mut i = 0;
    while i < encoded.len() {
        let c1 = ((encoded[i] as u32) << 16)
               | ((encoded[i+1] as u32) << 8)
               |  (encoded[i+2] as u32);
        let r1 = decode_word(c1);
        total_corrected += r1.corrected;
        if r1.uncorrectable { any_uncorrectable = true; }

        if out_idx < out_len { out[out_idx] = ((r1.msg >> 4) & 0xFF) as u8; out_idx += 1; }
        let b2_high = (r1.msg & 0x0F) << 4;

        if i + 3 < encoded.len() {
            let c2 = ((encoded[i+3] as u32) << 16)
                   | ((encoded[i+4] as u32) << 8)
                   |  (encoded[i+5] as u32);
            let r2 = decode_word(c2);
            total_corrected += r2.corrected;
            if r2.uncorrectable { any_uncorrectable = true; }
            if out_idx < out_len { out[out_idx] = (b2_high | (r2.msg >> 8) as u16) as u8; out_idx += 1; }
            if out_idx < out_len { out[out_idx] = (r2.msg & 0xFF) as u8; out_idx += 1; }
        } else {
            if out_idx < out_len { out[out_idx] = b2_high as u8; out_idx += 1; }
        }
        i += 6;
    }

    Ok(GolayDecoded {
        data: out[..out_idx].to_vec(),
        corrected: total_corrected,
        uncorrectable: any_uncorrectable,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_roundtrip() {
        let data = b"HelloTTT123456";
        let encoded = golay_encode(data);
        let decoded = golay_decode(&encoded).unwrap();
        assert_eq!(&decoded.data[..data.len()], data);
        assert!(!decoded.uncorrectable);
    }
    #[test]
    fn test_error_correction() {
        let data = b"TestData";
        let mut encoded = golay_encode(data);
        encoded[1] ^= 0b00000111; // flip 3 bits
        let decoded = golay_decode(&encoded).unwrap();
        assert_eq!(&decoded.data[..data.len()], data);
    }
}
